docker/pulsar/Dockerfile

```diff
@@ -26,82 +26,83 @@ ADD ${PULSAR_TARBALL} /
-     && apt-get update \
-     && apt-get -y dist-upgrade \
-     && apt-get -y install netcat dnsutils less procps iputils-ping \
-                 curl ca-certificates wget apt-transport-https \
-     && apt-get -y install --no-install-recommends python3 python3-kazoo python3-pip
-
-# Install Eclipse Temurin Package
-RUN mkdir -p /etc/apt/keyrings \
-     && wget -O - https://packages.adoptium.net/artifactory/api/gpg/key/public | tee /etc/apt/keyrings/adoptium.asc \
-     && echo "deb [signed-by=/etc/apt/keyrings/adoptium.asc] https://packages.adoptium.net/artifactory/deb $(awk -F= '/^VERSION_CODENAME/{print$2}' /etc/os-release) main" | tee /etc/apt/sources.list.d/adoptium.list \
-     && apt-get update \
-     && apt-get -y dist-upgrade \
-     && apt-get -y install temurin-${JDK_MAJOR_VERSION:-17}-jdk \
-     && export ARCH=$(uname -m | sed -r 's/aarch64/arm64/g' |  awk '!/arm64/{$0="amd64"}1') \
-     && echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/temurin-${JDK_MAJOR_VERSION:-17}-jdk-$ARCH/conf/security/java.security \
-     && echo networkaddress.cache.negative.ttl=1 >> /usr/lib/jvm/temurin-${JDK_MAJOR_VERSION:-17}-jdk-$ARCH/conf/security/java.security \
-
-# Cleanup apt
-RUN apt-get -y --purge autoremove \
-     && apt-get autoclean \
-     && apt-get clean \
-     && rm -rf /var/lib/apt/lists/*
-
-RUN pip3 install pyyaml==6.0.1
-
-# Pulsar currently writes to the below directories, assuming the default configuration.
-# Note that number 4 is the reason that pulsar components need write access to the /pulsar directory.
-# 1. /pulsar/data - both bookkeepers and zookeepers use this directory
-# 2. /pulsar/logs - function workers write to this directory and pulsar-admin initializes this directory
-# 3. /pulsar/download - functions write to this directory
-# 4. /pulsar - hadoop writes to this directory
-RUN mkdir /pulsar && chmod g+w /pulsar
+RUN apk add --no-cache \
+            bash \
+            python3 \
+            py3-pip \
+            ca-certificates
+
+RUN  wget -O /etc/apk/keys/amazoncorretto.rsa.pub https://apk.corretto.aws/amazoncorretto.rsa.pub && \
+    echo "https://apk.corretto.aws" >> /etc/apk/repositories && \
+    apk add --no-cache amazon-corretto-21 binutils && \
+    # Use JLink to create a slimmer JDK distribution (see: https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/)
+    # This still includes all JDK modules, though in the future we could compile a list of required modules
+    /usr/lib/jvm/default-jvm/bin/jlink --add-modules ALL-MODULE-PATH --compress zip-9 --no-man-pages --no-header-files --strip-debug --output /opt/corretto-slim && \
+    apk del binutils amazon-corretto-21 && \
+    mkdir -p /usr/lib/jvm/ && \
+    mv /opt/corretto-slim /usr/lib/jvm/java-21-amazon-corretto && \
+    ln -sfn /usr/lib/jvm/java-21-amazon-corretto /usr/lib/jvm/default-jvm
+
+ENV JAVA_HOME=/usr/lib/jvm/default-jvm
+
+RUN echo networkaddress.cache.ttl=1 >> /usr/lib/jvm/default-jvm/conf/security/java.security
```

Comment:
The negative cache TTL config seems to be missing?

Reply:
It is missing also in the current image. Honestly, I think we shouldn't be touching this anymore since we're anyway using Netty for all of ours critical DNS resolutions, so this doesn't even apply.

Reply:
I found it in the existing image here:
https://github.com/apache/pulsar/blob/7e73967f9daf368562fa0318d4ecb62272d72174/docker/pulsar/Dockerfile#L75

It is necessary. There's a reported issue #21206 and it was addressed by #21207.
Netty DNS settings will be also configured by using these settings with code in https://github.com/apache/pulsar/blob/7e73967f9daf368562fa0318d4ecb62272d72174/pulsar-common/src/main/java/org/apache/pulsar/common/util/netty/DnsResolverUtil.java#L42-L48
That's why it is matters also when using Netty.

Reply:
👍 I completely missed the negative cache line

Reply:
Fixed